main/wallet.js

```diff
@@ -0,0 +1,435 @@
+        }
+      }
+    }
+  `
+  const variables = { cid }
+  /** @type {{stateReplay: GQLStateReplay}} */
+  const { stateReplay } = await request(url, query, variables)
+  return stateReplay
+}
+
+/**
+ * @param {number} height
+ * @returns Promise<GQLTipset>
+ */
+async function getTipset (height) {
+  const query = gql`
+    query Tipset($height: Uint64!) {
+      tipset(height: $height) {
+        minTimestamp
+      }
+    }
+  `
+  const variables = { height }
+  const { tipset } = await request(url, query, variables)
+  return tipset
+}
+
+/**
+ * @param {string} address
+ * @returns Promise<GQLMessage[]>
+ */
+async function getMessages (address) {
+  const query = gql`
+    query Messages($address: String!, $limit: Int!, $offset: Int!) {
+      messages(address: $address, limit: $limit, offset: $offset) {
+        cid
+        to {
+          robust
+        }
+        from {
+          robust
+        }
+        nonce
+        height
+        method
+        params
+        value
+      }
+    }
+  `
+  const variables = {
+    address,
+    limit: 100,
+    offset: 0
+  }
+  /** @type {{messages: GQLMessage[]}} */
+  const { messages = [] } = await request(url, query, variables)
+  return messages
+}
+
+/**
+ * @returns {Promise<void>}
+ */
+async function updateTransactions () {
+  console.log(new Date(), 'updateTransactions')
+
+  // Load messages
+  const messages = await getMessages(address)
+
+  // Convert messages to transactions (loading)
+  /** @type {FILTransactionLoading[]} */
+  const transactionsLoading = messages.map(message => {
+    return {
+      height: message.height,
+      hash: message.cid,
+      outgoing: message.from.robust === address,
+      amount: new FilecoinNumber(message.value, 'attofil').toFil(),
+      address: message.from.robust === address
+        ? message.to.robust
+        : message.from.robust
+    }
+  })
+
+  /** @type {(FILTransaction|FILTransactionProcessing)[]} */
+  const updatedTransactions = []
+
+  for (const transactionProcessing of transactionsLoading) {
+    // Find matching transaction
+    const tx = transactions.find(tx => tx.hash === transactionProcessing.hash)
+
+    // Complete already loaded data
+    // Keep references alive by prefering `tx`
+    const transaction = tx || transactionProcessing
+
+    if (!transaction.timestamp && transaction.height) {
+      transaction.timestamp =
+        (await getTipset(transaction.height)).minTimestamp * 1000
+    }
+
+    if (
+      transaction.hash &&
+      (!transaction.status || transaction.status === 'processing') &&
+      !stateReplaysBeingFetched.has(transaction.hash)
+    ) {
+      const { hash } = transaction
+      stateReplaysBeingFetched.add(hash)
+      transaction.status = 'processing'
+      ;(async () => {
+        while (true) {
+          try {
+            const stateReplay = await getStateReplay(hash)
+            transaction.status = stateReplay.receipt.exitCode === 0
+              ? 'sent'
```

Comment:
I think the status `sent` can be confusing when it's shown for incoming transactions (even though they are unlikely).

Can we find a better name for this status? How about `confirmed` or `committed`?

Reply:
This is a good point! I changed it to `succeeded`, because I think a failed transaction can also be confirmed